### ISSUE TITLE
[Question] How to register a custom element?

### ISSUE THREAD

**Issue by MichaelRHaddad** [thingsboard/thingsboard#8683]
**Component**

 * Generic

**Description**
When I try to register a custom element, I get an error:
![image](https://github.com/thingsboard/thingsboard/assets/41021500/854cff72-3cec-4775-aa32-c24f9b7c6b74)

How to fix that? Changing the custom element's name doesn't help.

**Environment**

 * OS:  Win 10
 * ThingsBoard: 3.4.3 Pro
 * Browser: Firefox

**Comment by MichaelRHaddad** [comment]
@devaskim maybe you could help?

**Comment by devaskim** [comment]
Why I?)

P.S.: here is [similar SO topic ](https://stackoverflow.com/questions/74823213/navigating-back-causes-change-detection-to-break-in-angular-elements)

**Comment by MichaelRHaddad** [comment]
> Why I?)
> 
> P.S.: here is [similar SO topic ](https://stackoverflow.com/questions/74823213/navigating-back-causes-change-detection-to-break-in-angular-elements)

Well, you were very helpful once, so I thought I would tag you :)
Anyway... my question is not about Angular elements, but about regular [custom elements](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements). Would be grateful for your input!

**Comment by devaskim** [comment]
Honestly, I have never heard about this feature(((

**Comment by MichaelRHaddad** [comment]
@ashvayka help, please?

**Comment by devaskim** [comment]
Could you tell more about the final goal? Maybe it might be solved without custom elements.

**Comment by MichaelRHaddad** [comment]
> Could you tell more about the final goal? Maybe it might be solved without custom elements.

Well, I simply want nice-looking toggles instead of checkboxes:
![image](https://github.com/thingsboard/thingsboard/assets/41021500/0a34e768-f4a8-41eb-8f42-d8e38b712a71)

The HTML for each toggle is this:
```
<div class="toggle">
    <input type="checkbox" />
    <div class="tint"></div>
    <div class="slider"></div>
</div>
```
But when I have dozens or hundreds of them, the HTML gets very messy and polluted. Using custom elements allows me to encapsulate all the mess in a [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM) of a custom element:
```
<my-toggle></my-toggle>
````

I eventually simply used the `div`s version, but it is disgusting. This example is minor, but in the very near future I would need to embed some core custom elements and this will be a real problem for me.

**Comment by devaskim** [comment]
I see. How about Angular[ mat-slide-toggle](https://material.angular.io/components/slide-toggle/overview)?

**Comment by devdoob** [comment]
Why dont you use edault angular material [Slide Toggle](https://material.angular.io/components/slide-toggle/overview)? and apply styles to it

**Comment by MichaelRHaddad** [comment]
Truth be told, I do not know Angular, but working on it :)
I will try your suggestion.

**Comment by driesva** [comment]
It looks like your custom element is getting defined more than once. Maybe the HTML is loaded more than once?

Not sure if it will work, but you might try to define it conditionally:

```
    if (!customElements.get('bla-bla')) {
        customElements.define('bla-bla', BlaBla);
    }
```

See https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/get

**Comment by devaskim** [comment]
@MichaelRHaddad @Civilduino So here it is)

![Скриншот 30-01-2024 165700](https://github.com/thingsboard/thingsboard/assets/86909129/0fceea13-794a-46cf-95c6-c28ad5569ea3)

**Comment by MichaelRHaddad** [comment]
@devaskim 
Thank you. Could you please elaborate? What am I seeing here?

**Comment by devaskim** [comment]
You see a working example of a custom HTML element about which you asked me more than half year ago (look at your question in the begging of this issue).

**Comment by MichaelRHaddad** [comment]
@devaskim Thank you. I did understand that :) Could you please elaborate on the code itself? For instance, what exactly is this `customElementInited` field?

**Comment by devaskim** [comment]
> What am I seeing here?
> You see a working example

You got what I asked )))

My main idea is to postpone processing of widget's HTML template until we add custom elements. It is easy to implement with Angular features: `ngIf` and ability to provide widget's HTML template with custom data from JS code. The latter is done through storing data to `self.ctx.$scope` object. Then Angular does its magic every time those data is being changed.

So I added my variable `customElementInited ` to `$scope` and set it to `true` only after I add custom element. Then Angular receives this data update and re-process HTML template.

**Comment by Ultrazombie** [comment]
Closed due to inactivity
